agent/engine/docker_image_manager.go

```diff
@@ -303,21 +307,134 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
 	imageManager.updateLock.Lock()
 	defer imageManager.updateLock.Unlock()
 
-	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates())
+	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates(), ctx)
 	for i := 0; i < imageManager.numImagesToDelete; i++ {
 		err := imageManager.removeLeastRecentlyUsedImage(ctx)
 		if err != nil {
 			seelog.Infof("End of eligible images for deletion: %v; Still have %d image states being managed", err, len(imageManager.getAllImageStates()))
 			break
 		}
 	}
+	if imageManager.deleteNonECSImagesEnable {
+		// remove nonecs containers
+		imageManager.removeNonECSContainers(ctx)
+		// remove nonecs images
+		imageManager.removeNonECSImages(ctx)
+	}
+}
+
+func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Context) {
+	nonECSContainerIds := imageManager.nonECSContainersIds(ctx)
+	var nonECSContainerRemoveAvaiableIds []string
+	for _, id := range nonECSContainerIds {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if (response.State.Status == "exited") && (time.Now().Sub(response.State.FinishedAt) > imageManager.taskCleanupWaitDuration) {
+			nonECSContainerRemoveAvaiableIds = append(nonECSContainerRemoveAvaiableIds, id)
+		}
+	}
+	for _, id := range nonECSContainerRemoveAvaiableIds {
+		imageManager.client.RemoveContainer(ctx, id, dockerclient.RemoveContainerTimeout)
```

Comment:
we should not be removing all the dead containers here, since this may slow down docker. we should be doing something similar(certain count) to image cleanup

Reply:
yeah, will set the certain number in each cleanup period